counsel.el

```diff
@@ -6379,7 +6385,44 @@ sub-directories that builds may be invoked in."
   (let ((srcdir (counsel--compile-root)))
     (when (directory-files (or blddir srcdir) nil
                            counsel-compile-make-pattern t)
-      (counsel--compile-get-make-targets srcdir blddir))))
+      (counsel--compile-get-make-targets
+       'counsel-compile--probe-make-targets srcdir blddir))))
+
+
+(defun counsel-compile--probe-make-help (dir)
+  "Return a list of Make targets based on help for DIR.
+
+It is quite common for a 'make help' invocation to return a human
+readable list of targets.  Often common targets are marked with a
+leading asterisk.  The exact search pattern is controlled by
+`counsel-compile-help-pattern'."
+  (let ((default-directory dir)
+        (primary-targets nil)
+        (targets nil))
+    ;; only proceed if the help target exists
+    (when (= (call-process "make" nil t nil "-q" "help") 1)
+      (with-temp-buffer
+        (when (= (call-process "make" nil t nil "help") 0)
+          (goto-char (point-min))
+          (while (re-search-forward counsel-compile-help-pattern nil t)
+            (let ((text (match-string-no-properties 2)))
+            (if (match-beginning 1)
+                (setq primary-targets (nconc primary-targets (list text)))
+              (setq targets (nconc targets (list text))))))))
+      (append (sort primary-targets #'string-lessp)
+              (sort targets #'string-lessp)))))
```

Comment:
It is almost always more efficient to do something like:

```el
(let (elems)
  (while ...
    (push elem elems))
  (nreverse elems))
```

instead of appending each element to the list each time.

Since you are destructively sorting the lists at the end, however, the order of elements does not matter, so you can just write something like:

```el
(let (primary-targets targets)
  ...
  (push text (if (match-beginning 1) primary-targets targets))
  ...
  (nconc (sort primary-targets #'string-lessp)
         (sort targets #'string-lessp)))
```

Reply:
Hmm I've fixed it up as suggested but I also needed to add env support (for example the linux help output is affected by the compile environment). I wonder if this needs a bit of refactoring?

    (defun counsel-compile--probe-make-help (dir)
      "Return a list of Make targets based on help for `DIR'.

    It is quite common for a 'make help' invocation to return a human
    readable list of targets.  Often common targets are marked with a
    leading asterisk.  The exact search pattern is controlled by
    `counsel-compile-help-pattern'."
      (let* ((default-directory dir)
             (c-env (split-string
                     (substring-no-properties
                      (mapconcat #'identity counsel-compile-env " "))))
             (p-res (apply #'call-process "make" nil nil nil "-q" "help" c-env)))
        ;; only proceed if the help target exists
        (when (and (numberp p-res) (= p-res 1))
          (with-temp-buffer
            (let (primary-targets targets)
              (setq p-res
                    (apply #'call-process "make" nil t nil "help" c-env))
              (when (and (numberp p-res) (= p-res 0))
                (goto-char (point-min))
                (while (re-search-forward counsel-compile-help-pattern nil t)
                  (let ((text (match-string-no-properties 2)))
                    (push text
                          (if (string= "*" (match-string-no-properties 1))
                              primary-targets targets)))))
              (nconc (sort primary-targets #'string-lessp)
                     (sort targets #'string-lessp)))))))

Reply:
> I wonder if this needs a bit of refactoring?

Let me know if my comments address your concern.

Reply:
TIL eql, handy. Looking good

Reply:
> TIL eql

For fixnums you could equivalently use `eq`, but in general this is considered passé and `eql` is preferred instead, especially since the latter works with bignums in Emacs 27 and above:

```el
emacs-version
;; => "27.0.91"

(eq (1+ most-positive-fixnum)
    (1+ most-positive-fixnum))
;; => nil

(eql (1+ most-positive-fixnum)
     (1+ most-positive-fixnum))
;; => t
```